url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/WorkingMemory.java

```diff
@@ -0,0 +1,268 @@
+ * primarily by the ACS, but it can be used to facilitate interactions between the ACS and the NACS.
+ * <p>
+ * Working memory can contain any type of chunk. Also, unlike a goal, the dimensions of a working memory chunk 
+ * can have the same IDs as the dimensions in the sensory information space. This equates to "remembering" or
+ * "visualizing" something within the sensory information space and recalling it for later use.
+ * <p>
+ * Currently this is simply a container for the working memory slots. Any manipulations on the chunks 
+ * in working memory must be manually specified.
+ * <p>
+ * It is EXTREMELY important that you only manipulate the working memory using the add and remove
+ * methods that have been specifically defined by this class.
+ * <p>
+ * This class contains both global (static) and local constants. The default is to use the local 
+ * constants. If you want to change any of the global constants, you need to do so before any
+ * instances of this class are initialized.
+ * @version 6.0.6
+ * @author Nick Wilson
+ */
+public class WorkingMemory extends AbstractIntermediateModule <WorkingMemory.InfoStored, GenericChunkCollection>{
+	private static final long serialVersionUID = 3109928395636005779L;
+	
+	public enum InfoStored {TO_NACS, FROM_NACS, GENERIC_CHUNKS};
+	/**The number of slots in working memory.*/
+	public static int GLOBAL_CAPACITY = 7;
+	/**The number of slots in working memory.*/
+	public int CAPACITY = GLOBAL_CAPACITY;
+	
+	/**
+	 * Initializes the working memory. During initialization this instance of the working
+	 * memory will attach itself to the CLARION agent you specify.
+	 * @param Agent The agent to which the working memory is being attached.
+	 */
+	public WorkingMemory (CLARION Agent) 
+	{
+		super(Agent);
+	}
+	
+	/**
+	 * Adds a chunk to working memory. If you are trying to add a chunk but the working memory
+	 * is full, this method will throw an exception.
+	 * @param C The chunk to add.
+	 * @param TimeStamp The current time stamp.
+	 * @param The type of chunk being placed in working memory.
+	 * @throws FullContainerException If working memory is full.
+	 */
+	public void add (AbstractChunk C, Long TimeStamp, InfoStored type) throws FullContainerException
+	{
+		if(size() < CAPACITY)
+		{
+			EnumMap <InfoStored, GenericChunkCollection> e;
+			if(!containsKey(TimeStamp))
+				e = new EnumMap <InfoStored, GenericChunkCollection> (InfoStored.class);
+			else
+				e = super.get((Object)TimeStamp);
+			switch(type)
+			{
+			case TO_NACS:
+				if(!e.containsKey(InfoStored.TO_NACS))
+				{
+					GenericChunkCollection cc = new GenericChunkCollection ();
+					cc.put(C.getID(), C);
+					e.put(InfoStored.TO_NACS, cc);
+				}
+				else
+				{
+					GenericChunkCollection cc = e.get(InfoStored.TO_NACS);
+					cc.remove(C.getID());
+					cc.put(C.getID(), C);
+					e.put(InfoStored.TO_NACS, cc);
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 76:88, 91:103, 106:118.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.